File: src/types/websocket.ts
import type { Coordinates, HazardDetection, RobotHealth, RobotTask, VictimDetection } from "./domain";

// Discriminated union describing every inbound WS message.
// Matches backend channels: telemetry, robot_status, map_update, detection, mission_progress

export interface TelemetryMessage {
  type: "telemetry";
  robotId: string;
  timestamp: string;
  battery: number;
  speed: number;
  temperature: number;
  signal: number;
  position: Coordinates;
}

export interface RobotStatusMessage {
  type: "robot_status";
  robotId: string;
  health: RobotHealth;
  task: RobotTask;
  connection: "live" | "degraded" | "stale" | "offline";
  timestamp: string;
}

export interface MapUpdateMessage {
  type: "map_update";
  robotId: string;
  position: Coordinates;
  trail: Coordinates[];
  timestamp: string;
}

export interface DetectionMessage {
  type: "detection";
  kind: "victim" | "hazard";
  payload: VictimDetection | HazardDetection;
}

export interface MissionProgressMessage {
  type: "mission_progress";
  missionId: string;
  areaCoverage: number;
  robotsActive: number;
  timestamp: string;
}

export type WsInboundMessage =
  | TelemetryMessage
  | RobotStatusMessage
  | MapUpdateMessage
  | DetectionMessage
  | MissionProgressMessage;

export type WsConnectionState = "connecting" | "open" | "reconnecting" | "closed";

File: backend/app/schemas/websocket.py
"""
WebSocket message schemas matching websocket.ts discriminated union.
"""

from pydantic import BaseModel

from app.schemas.robot import Coordinates


class TelemetryWsMessage(BaseModel):
    type: str = "telemetry"
    robotId: str
    timestamp: str
    battery: float
    speed: float
    temperature: float
    signal: float
    position: Coordinates


class RobotStatusWsMessage(BaseModel):
    type: str = "robot_status"
    robotId: str
    health: str
    task: str
    connection: str
    timestamp: str


class MapUpdateWsMessage(BaseModel):
    type: str = "map_update"
    robotId: str
    position: Coordinates
    trail: list[Coordinates]
    timestamp: str


class DetectionWsMessage(BaseModel):
    type: str = "detection"
    kind: str  # "victim" | "hazard"
    payload: dict


class MissionProgressWsMessage(BaseModel):
    type: str = "mission_progress"
    missionId: str
    areaCoverage: float
    robotsActive: int
    timestamp: str
